tests/ui/features/bootstrap/TextEditorContext.php

```diff
@@ -0,0 +1,166 @@
+	private $textEditorPage;
+	private $featureContext;
+	private $filesContext;
+
+	/**
+	 * TextEditorContext constructor.
+	 *
+	 * @param TextEditorPage $textEditorPage
+	 */
+	public function __construct(TextEditorPage $textEditorPage) {
+		$this->textEditorPage = $textEditorPage;
+	}
+
+	/**
+	 * @When /^I create a text file with the name ((?:'[^']*')|(?:"[^"]*"))( without changing the default file extension|)$/
+	 *
+	 * @param string $name
+	 * @param string $useDefaultFileExtension
+	 * @return void
+	 */
+	public function createATextFileWithTheName(
+		$name,
+		$useDefaultFileExtension = ''
+	) {
+		// The capturing group of the regex always includes the quotes at each
+		// end of the captured string, so trim them.
+		$name = trim($name, $name[0]);
+		$this->textEditorPage->createTextFile(
+			$this->getSession(),
+			$name,
+			strlen($useDefaultFileExtension) ? true : false
+		);
+		$this->textEditorPage->waitTillEditorIsLoaded();
+	}
+
+
+	/**
+	 * @Then near the new text file box a tooltip with the text :toolTipText should be displayed
+	 */
+	public function nearTheNewTextFileBoxATooltipShouldBeDisplayed($toolTipText)
+	{
+		PHPUnit_Framework_Assert::assertEquals($toolTipText,
+			$this->textEditorPage->getTooltipOfNewTextFileBox()
+		);
+	}
+
+	/**
+	 * @When I input :text in the text area
+	 * @param string $text
+	 * @return void
+	 */
+	public function iInputTextInTheTextArea($text) {
+		$this->textEditorPage->typeIntoTextFile(
+			$this->getSession(),
+			$text
+		);
+	}
+
+	/**
+	 * @When I input the following text in the text area:
+	 * @param PyStringNode $multiLineText
+	 * @return void
+	 */
+	public function iInputTheFollowingInTheTextArea(PyStringNode $multiLineText) {
+		$this->textEditorPage->typeIntoTextFile(
+			$this->getSession(),
+			$multiLineText->getRaw()
+		);
+	}
+
+	/**
+	 * @Then there is/are :number line(s) of text
+	 * @param int $number
+	 * @return void
+	 */
+	public function thereAreLinesOfText($number) {
+		PHPUnit_Framework_Assert::assertEquals(
+			$number,
+			count($this->textEditorPage->textFileContent())
+		);
+	}
+
+	/**
+	 * @Then line :number of the text is :text
+	 * @param int $number
+	 * @param string $text
+	 * @return void
+	 */
+	public function lineOfTheTextIs($number, $text) {
+		$lineIndex = $number - 1;
+		$textFileContent = $this->textEditorPage->textFileContent();
+		PHPUnit_Framework_Assert::assertEquals(
+			$text,
+			$textFileContent[$lineIndex]
+		);
+	}
+	/**
+	 * @When I close the text editor
+	 * @return void
+	 */
+	public function iCloseTheTextEditor() {
+		// Issue https://github.com/owncloud/files_texteditor/issues/205
+		// text file auto-save happens asynchronously at intervals
+		// give auto-save a chance to have saved recent changes
+		// before we close the text editor.
+		// Note: in testing with a 2 second sleep there were still some fails
+		// so it needs quite a long sleep to ensure the latest content has
+		// been saved.
+		sleep(5);
```

Comment:
could we use the waitTillAjaxStartAndFinish function here?

Reply:
I will play with it. I think the ACE code just saves at timed intervals when there are changes (i.e. save does not trigger on every input keystroke) - so it is possible to finish typing and nothing will be happening then some (short) time later "save" will wake up.

But I will check.